File: Repositorio/engenharia-de-software-com-ia-aplicada-main/modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-03-fine-tuning-via-api/dolly-dataset-real-starter.js
/**
 * Ahirton Lopes · Fine-Tuning Toolkit
 * Extra - Dataset real alternativo, parte 1/2 (companion do Módulo 3.2 em
 * diante, mas conceitualmente equivalente ao Módulo 2.2 - preparação de
 * dataset, não upload/job)
 *
 * Ferramenta: carrega o databricks-dolly-15k (real, CC-BY-SA-3.0), mapeia pro
 * schema canônico do Módulo 2.1 (instrucao/entrada/saida/metadata), e roda
 * o mesmo tipo de dedup MinHash+LSH e balanceamento por temperatura do
 * Módulo 2.2 contra dado real, não sintético. Fica fisicamente na pasta do
 * Módulo 3 (junto com a parte 2/2, dolly-vertex-pipeline.js, que sobe e
 * treina de verdade) só por conveniência de não espalhar o extra em duas
 * pastas - o conteúdo em si é 100% Módulo 2.2: dedup e balanceamento, nada
 * de upload ou job.
 *
 * Pré-requisito: baixar o dataset uma vez (13MB, ~15 mil linhas JSONL):
 *   curl -L "https://huggingface.co/datasets/databricks/databricks-dolly-15k/resolve/main/databricks-dolly-15k.jsonl" -o databricks-dolly-15k.jsonl
 *
 * Nota de honestidade: das 8 categorias do Dolly-15k, só 3 têm o campo
 * "context" preenchido (closed_qa, information_extraction, summarization) -
 * são as únicas com o mesmo formato "documento de entrada -> saída
 * específica" que o schema canônico exige. As outras 5 (open_qa,
 * general_qa, classification, brainstorming, creative_writing) não têm
 * documento de entrada real, então não servem pra esse pipeline: usar só
 * as 3 compatíveis não é recorte arbitrário, é o que o próprio formato do
 * dado permite.
 *
 * Segunda nota de honestidade: as funções `encontrarQuaseDuplicatasMinHashLSH`
 * e `limparEBalancear`, importadas de dataset-cleaning-balancing-tool.js, são
 * hardcoded pros dois casos desta disciplina ('amplitude-auto' e
 * 'amplitude-saude-empresarial') - não acionam em cima de um `metadata.caso`
 * novo. Balanceamento por temperatura (balancearPorTemperatura,
 * entropiaShannon, numeroEfetivoFontes) já é genérico e reusado direto daqui;
 * dedup usa as mesmas primitivas exportadas (assinaturaMinHash, bandingLSH,
 * similaridadeJaccardExata), só reimplementando o laço de orquestração de
 * forma genérica, ~15 linhas, em vez de patchear o arquivo original.
 */

const fs = require('fs');
const path = require('path');
const assert = require('assert');
const m2 = require('../modulo-02-preparacao-datasets/dataset-cleaning-balancing-tool.js');

const CATEGORIAS_COMPATIVEIS = ['information_extraction', 'closed_qa', 'summarization'];
const CASO = 'dolly-instruction-tuning';
const N_SHINGLE = 5;

/* --------------------------------------------------------------------------
 * 1. Carregar e mapear pro schema canônico
 * -------------------------------------------------------------------------- */

function carregarDolly(caminhoJsonl) {
  const bruto = fs.readFileSync(caminhoJsonl, 'utf-8').trim().split('\n');
  return bruto.map((linha) => JSON.parse(linha));
}

/** Mapeia um registro bruto do Dolly-15k pro schema canônico do Módulo 2.1. */
function paraSchemaCanonico(registroDolly, indice) {
  return {
    instrucao: registroDolly.instruction,
    entrada: registroDolly.context,
    saida: registroDolly.response,
    metadata: {
      caso: CASO,
      fonte: registroDolly.category,
      id: `dolly-${registroDolly.category}-${indice}`,
    },
  };
}

function filtrarCompativeis(registrosDolly) {
  return registrosDolly.filter(
    (r) => CATEGORIAS_COMPATIVEIS.includes(r.category) && r.context && r.context.trim() && r.response && r.response.trim()
  );
}

/* --------------------------------------------------------------------------
 * 2. Dedup genérico via MinHash+LSH (reusa as primitivas do Módulo 2.2,
 *    não o orquestrador hardcoded pros casos Amplitude)
 * -------------------------------------------------------------------------- */

function shinglesLocal(texto, n) {
  const palavras = texto.toLowerCase().replace(/\s+/g, ' ').trim().split(' ');
  const conjunto = new Set();
  for (let i = 0; i <= palavras.length - n; i++) conjunto.add(palavras.slice(i, i + n).join(' '));
  return conjunto;
}

/** contarPorFonte não é exportado por dataset-cleaning-balancing-tool.js; reimplementado aqui, mesma lógica. */
function contarPorFonteLocal(exemplos, caso) {
  const doCaso = exemplos.filter((e) => e.metadata.caso === caso);
  const contagem = {};
  for (const e of doCaso) contagem[e.metadata.fonte] = (contagem[e.metadata.fonte] || 0) + 1;
  return contagem;
}

/**
 * Comparar só `entrada` confunde pares de PERGUNTA DIFERENTE sobre o mesmo
 * trecho de contexto (Jaccard=1,0 na entrada) com duplicata real.
 * `textoParaDedup` concatena instrução+entrada antes de shinglar, então
 * duas perguntas diferentes sobre o mesmo texto-fonte já não colidem mais.
 */
function textoParaDedup(exemplo) {
  return `${exemplo.instrucao}\n${exemplo.entrada}`;
}

function dedupGenerico(exemplos, n = N_SHINGLE) {
  const coeficientes = m2.gerarCoeficientesHash(m2.MINHASH_K, 42);
  const assinaturas = exemplos.map((e) => m2.assinaturaMinHash(shinglesLocal(textoParaDedup(e), n), coeficientes));
  const candidatos = m2.bandingLSH(assinaturas, m2.LSH_BANDAS, m2.LSH_LINHAS);
  const pares = [];
  for (const chave of candidatos) {
    const [i, j] = chave.split('-').map(Number);
    const sim = m2.similaridadeJaccardExata(textoParaDedup(exemplos[i]), textoParaDedup(exemplos[j]), n);
    if (sim >= m2.LIMIAR_DUPLICATA) pares.push({ i, j, sim });
  }
  const totalForcaBruta = (exemplos.length * (exemplos.length - 1)) / 2;
  return { totalForcaBruta, candidatosLSH: candidatos.size, pares };
}

/**
 * Pipeline completo de preparo: roda dedup no dataset INTEIRO (não amostra)
 * e balanceia pro alvo pedido, pra que os números do model card sejam
 * reproduzíveis a partir do arquivo entregue. É exatamente o que gerou o
 * dataset do job real.
 */
function prepararDatasetCompleto(caminhoJsonl, alvoTotal) {
  const bruto = carregarDolly(caminhoJsonl);
  const compativeis = filtrarCompativeis(bruto);
  const mapeados = compativeis.map(paraSchemaCanonico);

  const dedup = dedupGenerico(mapeados, N_SHINGLE);
  const remover = new Set(dedup.pares.map((p) => p.j));
  const semDuplicatas = mapeados.filter((_, i) => !remover.has(i));

  const contagem = contarPorFonteLocal(semDuplicatas, CASO);
  const alocacao = m2.alocarComCapacidade(contagem, m2.ALPHA_TEMPERATURA, alvoTotal);
  const usados = {};
  const balanceado = [];
  for (const e of semDuplicatas) {
    const f = e.metadata.fonte;
    usados[f] = usados[f] || 0;
    if (usados[f] < alocacao[f]) { balanceado.push(e); usados[f] += 1; }
  }

  return {
    bruto, compativeis, mapeados, dedup,
    itensRemovidos: remover.size, semDuplicatas,
    contagem, alocacao, balanceado,
  };
}

/* --------------------------------------------------------------------------
 * Testes automatizados
 * -------------------------------------------------------------------------- */

let totalTestes = 0;
let testesComFalha = 0;

function testar(descricao, fn) {
  totalTestes += 1;
  try {
    fn();
    console.log(`  [OK] ${descricao}`);
  } catch (erro) {
    testesComFalha += 1;
    console.log(`  [FALHOU] ${descricao}`);
    console.log(`           ${erro.message}`);
  }
}

function rodarTestes(caminhoJsonl) {
  console.log('== Testes: carga e mapeamento pro schema canônico ==');
  const bruto = carregarDolly(caminhoJsonl);

  testar('dataset bruto tem 15011 registros (tamanho real confirmado do Dolly-15k)', () => {
    assert.equal(bruto.length, 15011);
  });

  testar('só 3 das 8 categorias têm campo context preenchido', () => {
    const porCategoria = {};
    for (const r of bruto) {
      porCategoria[r.category] = porCategoria[r.category] || { total: 0, comContexto: 0 };
      porCategoria[r.category].total += 1;
      if (r.context && r.context.trim()) porCategoria[r.category].comContexto += 1;
    }
    for (const cat of CATEGORIAS_COMPATIVEIS) {
      assert.ok(porCategoria[cat].comContexto === porCategoria[cat].total, `${cat} deveria ter 100% com contexto`);
    }
    const incompativeis = Object.keys(porCategoria).filter((c) => !CATEGORIAS_COMPATIVEIS.includes(c));
    for (const cat of incompativeis) {
      assert.equal(porCategoria[cat].comContexto, 0, `${cat} deveria ter 0% com contexto`);
    }
  });

  const compativeis = filtrarCompativeis(bruto);
  testar('filtro produz 4467 exemplos reais compatíveis com o schema canônico', () => {
    assert.equal(compativeis.length, 4467);
  });

  const mapeados = compativeis.map(paraSchemaCanonico);
  testar('todo exemplo mapeado tem os quatro campos do schema canônico preenchidos', () => {
    for (const e of mapeados) {
      assert.ok(e.instrucao && e.entrada && e.saida, 'campo vazio encontrado');
      assert.ok(e.metadata.caso && e.metadata.fonte, 'metadata incompleto');
    }
  });

  testar('distribuição por fonte bate com a contagem real do dataset', () => {
    const porFonte = {};
    for (const e of mapeados) porFonte[e.metadata.fonte] = (porFonte[e.metadata.fonte] || 0) + 1;
    assert.equal(porFonte.closed_qa, 1773);
    assert.equal(porFonte.information_extraction, 1506);
    assert.equal(porFonte.summarization, 1188);
  });

  console.log('\n== Testes: dedup genérico, comparando instrução+entrada ==');

  testar('entrada sozinha teria Jaccard 1,0 (é POR ISSO que comparar só entrada é errado aqui)', () => {
    const a = mapeados.find((e) => e.instrucao === 'What caused the Global Financial Crises?');
    const b = mapeados.find((e) => e.instrucao === 'What caused the 2007-2008 financial crisis?');
    assert.ok(a && b, 'par de teste não encontrado no dataset baixado');
    assert.equal(m2.similaridadeJaccardExata(a.entrada, b.entrada, N_SHINGLE), 1);
  });

  testar('com instrução+entrada, esse mesmo par NÃO é mais Jaccard 1,0 (fix funcionando)', () => {
    const a = mapeados.find((e) => e.instrucao === 'What caused the Global Financial Crises?');
    const b = mapeados.find((e) => e.instrucao === 'What caused the 2007-2008 financial crisis?');
    assert.ok(m2.similaridadeJaccardExata(textoParaDedup(a), textoParaDedup(b), N_SHINGLE) < 1);
  });

  console.log('\n== Testes: pipeline completo (dataset inteiro, não amostra) ==');

  const completo = prepararDatasetCompleto(caminhoJsonl, 200);
  testar('roda contra os 4.467 exemplos inteiros, não uma amostra', () => {
    assert.equal(completo.mapeados.length, 4467);
  });

  testar('LSH reduz a busca de força-bruta drasticamente também em dado real', () => {
    assert.ok(completo.dedup.candidatosLSH < completo.dedup.totalForcaBruta * 0.01, 'LSH não reduziu o suficiente');
  });

  testar('dedup real encontra pelo menos 1 par genuíno (mesma instrução E mesmo contexto)', () => {
    assert.ok(completo.dedup.pares.length > 0, 'nenhuma quase-duplicata real encontrada');
  });

  testar('itens removidos é a contagem de índices ÚNICOS, não o total de pares (podem se sobrepor)', () => {
    assert.ok(completo.itensRemovidos <= completo.dedup.pares.length, 'itens removidos não pode exceder pares confirmados');
    assert.equal(completo.semDuplicatas.length, completo.mapeados.length - completo.itensRemovidos);
  });

  console.log('\n== Testes: balanceamento por temperatura (reusado sem modificação) ==');

  testar('alocação final soma exatamente o alvo pedido (200)', () => {
    assert.equal(completo.balanceado.length, 200);
  });

  testar('alocação capacitada nunca excede a contagem real disponível por fonte', () => {
    for (const fonte of Object.keys(completo.alocacao)) {
      assert.ok(completo.alocacao[fonte] <= completo.contagem[fonte], `${fonte} excedeu a contagem real`);
    }
  });

  const distAntes = m2.distribuicaoDe(completo.contagem);
  const distDepois = m2.distribuicaoDe(completo.alocacao);
  testar('entropia sobe depois do balanceamento, igual no dado sintético do Módulo 2.2', () => {
    assert.ok(m2.entropiaShannon(distDepois) >= m2.entropiaShannon(distAntes));
  });

  console.log(`\n${totalTestes - testesComFalha}/${totalTestes} testes passaram.`);
  return { bruto, compativeis, mapeados, completo, distAntes, distDepois };
}

/* --------------------------------------------------------------------------
 * Demo
 * -------------------------------------------------------------------------- */

function rodarDemo(caminhoJsonl) {
  console.log('===== Extra: dataset real alternativo (Dolly-15k) rodando contra o pipeline do Módulo 2.2 =====\n');
  const r = rodarTestes(caminhoJsonl);
  const c = r.completo;

  console.log('\n--- Resumo ---');
  console.log(`Dataset bruto: ${r.bruto.length} exemplos reais (databricks-dolly-15k, CC-BY-SA-3.0)`);
  console.log(`Compatíveis com o schema canônico (têm contexto real): ${r.compativeis.length}`);
  console.log(`Distribuição por fonte antes do dedup: ${JSON.stringify(contarPorFonteLocal(r.mapeados, CASO))}`);
  console.log(`Dedup no dataset INTEIRO (comparando instrução+entrada): ${c.dedup.candidatosLSH} candidatos LSH de ${c.dedup.totalForcaBruta} pares força-bruta, ${c.dedup.pares.length} pares confirmados, ${c.itensRemovidos} itens únicos removidos -> ${c.semDuplicatas.length} restantes`);
  console.log(`Balanceado pro alvo: ${JSON.stringify(c.alocacao)}, total ${c.balanceado.length}`);
  console.log(`Entropia antes/depois do balanceamento: ${m2.entropiaShannon(r.distAntes).toFixed(4)} -> ${m2.entropiaShannon(r.distDepois).toFixed(4)} nats`);
  console.log('\nComparar só `entrada` no dedup confundiria pares de PERGUNTA DIFERENTE sobre o MESMO trecho de contexto (Jaccard=1,0 na entrada) com duplicata real - ex.: duas perguntas diferentes sobre a crise financeira de 2008, mesmo texto-fonte. Comparar instrução+entrada evita isso: esse par específico não é candidato a duplicata.');
}

module.exports = {
  carregarDolly,
  paraSchemaCanonico,
  filtrarCompativeis,
  dedupGenerico,
  textoParaDedup,
  prepararDatasetCompleto,
  contarPorFonteLocal,
  rodarTestes,
  CATEGORIAS_COMPATIVEIS,
  CASO,
};

if (require.main === module) {
  const caminho = process.argv[2] || path.join(__dirname, 'databricks-dolly-15k.jsonl');
  if (!fs.existsSync(caminho)) {
    console.error(`Arquivo não encontrado: ${caminho}`);
    console.error('Baixe primeiro: curl -L "https://huggingface.co/datasets/databricks/databricks-dolly-15k/resolve/main/databricks-dolly-15k.jsonl" -o databricks-dolly-15k.jsonl');
    process.exit(1);
  }
  rodarDemo(caminho);
}

/*
 * Ahirton Lopes · Fine-Tuning Toolkit - UNIPDS: Processamento de Dados e Fine-Tuning de Modelos
 * Prof. Ahirton Lopes, Ph.D. - GDE AI, Microsoft MVP, Senior Manager
 */

File: Repositorio/engenharia-de-software-com-ia-aplicada-main/modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-03-fine-tuning-via-api/dolly_dataset_real_starter.py
"""
Ahirton Lopes - Fine-Tuning Toolkit
Extra - Dataset real alternativo, parte 1/2 (companion do Modulo 3.2 em
diante, mas conceitualmente equivalente ao Modulo 2.2 - preparacao de
dataset, nao upload/job; referencia espelhada em Python)

Ferramenta: carrega o databricks-dolly-15k (real, CC-BY-SA-3.0), mapeia pro
schema canonico do Modulo 2.1 (instrucao/entrada/saida/metadata), e roda o
mesmo tipo de dedup MinHash+LSH e balanceamento por temperatura do
Modulo 2.2 contra dado real, nao sintetico. Fica fisicamente na pasta do
Modulo 3 (junto com a parte 2/2, dolly_vertex_pipeline.py, que sobe e
treina de verdade) so por conveniencia de nao espalhar o extra em duas
pastas - o conteudo em si e 100% Modulo 2.2: dedup e balanceamento, nada
de upload ou job.

Pre-requisito: baixar o dataset uma vez (13MB, ~15 mil linhas JSONL):
  curl -L "https://huggingface.co/datasets/databricks/databricks-dolly-15k/resolve/main/databricks-dolly-15k.jsonl" -o databricks-dolly-15k.jsonl

Nota de honestidade: das 8 categorias do Dolly-15k, so 3 tem o campo
"context" preenchido (closed_qa, information_extraction, summarization) -
sao as unicas com o mesmo formato "documento de entrada -> saida
especifica" que o schema canonico exige. As outras 5 (open_qa, general_qa,
classification, brainstorming, creative_writing) nao tem documento de
entrada real, entao nao servem pra esse pipeline: usar so as 3
compativeis nao e recorte arbitrario, e o que o proprio formato do dado
permite.

Segunda nota de honestidade: a funcao `encontrar_quase_duplicatas_minhash_lsh`
e `limpar_e_balancear`, importadas de dataset_cleaning_balancing_tool.py, sao
hardcoded pros dois casos desta disciplina ('amplitude-auto' e
'amplitude-saude-empresarial') - nao acionam em cima de um metadata["caso"]
novo. Balanceamento por temperatura (balancear_por_temperatura,
entropia_shannon, numero_efetivo_fontes, contar_por_fonte) ja e generico e
reusado direto daqui; dedup usa as mesmas primitivas exportadas (shingles,
assinatura_minhash, banding_lsh, similaridade_jaccard_exata), so
reimplementando o laco de orquestracao de forma generica, ~15 linhas, em
vez de patchear o arquivo original.
"""

import importlib.util
import json
import os
import sys

_AQUI = os.path.dirname(os.path.abspath(__file__))
_CAMINHO_M22 = os.path.join(_AQUI, "..", "modulo-02-preparacao-datasets", "dataset_cleaning_balancing_tool.py")
_spec = importlib.util.spec_from_file_location("dataset_cleaning_balancing_tool", _CAMINHO_M22)
m2 = importlib.util.module_from_spec(_spec)
_spec.loader.exec_module(m2)

CATEGORIAS_COMPATIVEIS = ["information_extraction", "closed_qa", "summarization"]
CASO = "dolly-instruction-tuning"
N_SHINGLE = 5


# -----------------------------------------------------------------------
# 1. Carregar e mapear pro schema canonico
# -----------------------------------------------------------------------

def carregar_dolly(caminho_jsonl):
    with open(caminho_jsonl, encoding="utf-8") as f:
        return [json.loads(linha) for linha in f if linha.strip()]


def para_schema_canonico(registro_dolly, indice):
    """Mapeia um registro bruto do Dolly-15k pro schema canonico do Modulo 2.1."""
    return {
        "instrucao": registro_dolly["instruction"],
        "entrada": registro_dolly["context"],
        "saida": registro_dolly["response"],
        "metadata": {
            "caso": CASO,
            "fonte": registro_dolly["category"],
            "id": f"dolly-{registro_dolly['category']}-{indice}",
        },
    }


def filtrar_compativeis(registros_dolly):
    return [
        r for r in registros_dolly
        if r["category"] in CATEGORIAS_COMPATIVEIS and r.get("context", "").strip() and r.get("response", "").strip()
    ]


# -----------------------------------------------------------------------
# 2. Dedup generico via MinHash+LSH (reusa as primitivas do Modulo 2.2,
#    nao o orquestrador hardcoded pros casos Amplitude)
# -----------------------------------------------------------------------

def texto_para_dedup(exemplo):
    """Comparar so `entrada` confunde pares de PERGUNTA DIFERENTE sobre o
    mesmo trecho de contexto (Jaccard=1,0 na entrada) com duplicata real.
    Concatenar instrucao+entrada evita isso."""
    return f"{exemplo['instrucao']}\n{exemplo['entrada']}"


def dedup_generico(exemplos, n=N_SHINGLE):
    coeficientes = m2.gerar_coeficientes_hash(m2.MINHASH_K, 42)
    assinaturas = [m2.assinatura_minhash(m2.shingles(texto_para_dedup(e), n), coeficientes) for e in exemplos]
    candidatos = m2.banding_lsh(assinaturas, m2.LSH_BANDAS, m2.LSH_LINHAS)
    pares = []
    for i, j in candidatos:
        sim = m2.similaridade_jaccard_exata(texto_para_dedup(exemplos[i]), texto_para_dedup(exemplos[j]), n)
        if sim >= m2.LIMIAR_DUPLICATA:
            pares.append({"i": i, "j": j, "sim": sim})
    total_forca_bruta = len(exemplos) * (len(exemplos) - 1) // 2
    return {"total_forca_bruta": total_forca_bruta, "candidatos_lsh": len(candidatos), "pares": pares}


def preparar_dataset_completo(caminho_jsonl, alvo_total):
    """Pipeline completo de preparo: roda dedup no dataset INTEIRO e balanceia
    pro alvo, pra que os numeros do model card sejam reproduziveis a partir
    do arquivo entregue."""
    bruto = carregar_dolly(caminho_jsonl)
    compativeis = filtrar_compativeis(bruto)
    mapeados = [para_schema_canonico(r, i) for i, r in enumerate(compativeis)]

    dedup = dedup_generico(mapeados, N_SHINGLE)
    remover = {p["j"] for p in dedup["pares"]}
    sem_duplicatas = [e for i, e in enumerate(mapeados) if i not in remover]

    contagem = m2.contar_por_fonte(sem_duplicatas, CASO)
    alocacao = m2.alocar_com_capacidade(contagem, m2.ALPHA_TEMPERATURA, alvo_total)
    usados = {}
    balanceado = []
    for e in sem_duplicatas:
        f = e["metadata"]["fonte"]
        usados[f] = usados.get(f, 0)
        if usados[f] < alocacao[f]:
            balanceado.append(e)
            usados[f] += 1

    return {
        "bruto": bruto, "compativeis": compativeis, "mapeados": mapeados, "dedup": dedup,
        "itens_removidos": len(remover), "sem_duplicatas": sem_duplicatas,
        "contagem": contagem, "alocacao": alocacao, "balanceado": balanceado,
    }


# -----------------------------------------------------------------------
# Testes automatizados
# -----------------------------------------------------------------------

_total_testes = 0
_testes_com_falha = 0


def _testar(descricao, fn):
    global _total_testes, _testes_com_falha
    _total_testes += 1
    try:
        fn()
        print(f"  [OK] {descricao}")
    except AssertionError as erro:
        _testes_com_falha += 1
        print(f"  [FALHOU] {descricao}")
        print(f"           {erro}")


def rodar_testes(caminho_jsonl):
    print("== Testes: carga e mapeamento pro schema canonico ==")
    bruto = carregar_dolly(caminho_jsonl)

    _testar("dataset bruto tem 15011 registros (tamanho real confirmado do Dolly-15k)", lambda: (
        _assert(len(bruto) == 15011)
    ))

    def _checar_categorias():
        por_categoria = {}
        for r in bruto:
            d = por_categoria.setdefault(r["category"], {"total": 0, "com_contexto": 0})
            d["total"] += 1
            if r.get("context", "").strip():
                d["com_contexto"] += 1
        for cat in CATEGORIAS_COMPATIVEIS:
            _assert(por_categoria[cat]["com_contexto"] == por_categoria[cat]["total"], f"{cat} deveria ter 100% com contexto")
        for cat in por_categoria:
            if cat not in CATEGORIAS_COMPATIVEIS:
                _assert(por_categoria[cat]["com_contexto"] == 0, f"{cat} deveria ter 0% com contexto")

    _testar("so 3 das 8 categorias tem campo context preenchido", _checar_categorias)

    compativeis = filtrar_compativeis(bruto)
    _testar("filtro produz 4467 exemplos reais compativeis com o schema canonico", lambda: (
        _assert(len(compativeis) == 4467)
    ))

    mapeados = [para_schema_canonico(r, i) for i, r in enumerate(compativeis)]

    def _checar_campos():
        for e in mapeados:
            _assert(e["instrucao"] and e["entrada"] and e["saida"], "campo vazio encontrado")
            _assert(e["metadata"]["caso"] and e["metadata"]["fonte"], "metadata incompleto")

    _testar("todo exemplo mapeado tem os quatro campos do schema canonico preenchidos", _checar_campos)

    def _checar_distribuicao():
        por_fonte = {}
        for e in mapeados:
            por_fonte[e["metadata"]["fonte"]] = por_fonte.get(e["metadata"]["fonte"], 0) + 1
        _assert(por_fonte["closed_qa"] == 1773)
        _assert(por_fonte["information_extraction"] == 1506)
        _assert(por_fonte["summarization"] == 1188)

    _testar("distribuicao por fonte bate com a contagem real do dataset", _checar_distribuicao)

    print("\n== Testes: dedup generico, comparando instrucao+entrada ==")

    def _checar_entrada_sozinha_seria_1():
        a = next(e for e in mapeados if e["instrucao"] == "What caused the Global Financial Crises?")
        b = next(e for e in mapeados if e["instrucao"] == "What caused the 2007-2008 financial crisis?")
        _assert(m2.similaridade_jaccard_exata(a["entrada"], b["entrada"], N_SHINGLE) == 1)

    _testar("entrada sozinha teria Jaccard 1,0 (e POR ISSO que comparar so entrada e errado aqui)", _checar_entrada_sozinha_seria_1)

    def _checar_fix_funciona():
        a = next(e for e in mapeados if e["instrucao"] == "What caused the Global Financial Crises?")
        b = next(e for e in mapeados if e["instrucao"] == "What caused the 2007-2008 financial crisis?")
        _assert(m2.similaridade_jaccard_exata(texto_para_dedup(a), texto_para_dedup(b), N_SHINGLE) < 1)

    _testar("com instrucao+entrada, esse mesmo par NAO e mais Jaccard 1,0 (fix funcionando)", _checar_fix_funciona)

    print("\n== Testes: pipeline completo (dataset inteiro, nao amostra) ==")

    completo = preparar_dataset_completo(caminho_jsonl, 200)

    _testar("roda contra os 4.467 exemplos inteiros, nao uma amostra", lambda: (
        _assert(len(completo["mapeados"]) == 4467)
    ))
    _testar("LSH reduz a busca de forca-bruta drasticamente tambem em dado real", lambda: (
        _assert(completo["dedup"]["candidatos_lsh"] < completo["dedup"]["total_forca_bruta"] * 0.01, "LSH nao reduziu o suficiente")
    ))
    _testar("dedup real encontra pelo menos 1 par genuino (mesma instrucao E mesmo contexto)", lambda: (
        _assert(len(completo["dedup"]["pares"]) > 0, "nenhuma quase-duplicata real encontrada")
    ))

    def _checar_itens_removidos():
        _assert(completo["itens_removidos"] <= len(completo["dedup"]["pares"]))
        _assert(len(completo["sem_duplicatas"]) == len(completo["mapeados"]) - completo["itens_removidos"])

    _testar("itens removidos e a contagem de indices UNICOS, nao o total de pares", _checar_itens_removidos)

    print("\n== Testes: balanceamento por temperatura (reusado sem modificacao) ==")

    _testar("alocacao final soma exatamente o alvo pedido (200)", lambda: (
        _assert(len(completo["balanceado"]) == 200)
    ))

    def _checar_capacidade():
        for fonte, n in completo["alocacao"].items():
            _assert(n <= completo["contagem"][fonte], f"{fonte} excedeu a contagem real")

    _testar("alocacao capacitada nunca excede a contagem real disponivel por fonte", _checar_capacidade)

    dist_antes = m2.distribuicao_de(completo["contagem"])
    dist_depois = m2.distribuicao_de(completo["alocacao"])
    _testar("entropia sobe depois do balanceamento, igual no dado sintetico do Modulo 2.2", lambda: (
        _assert(m2.entropia_shannon(dist_depois) >= m2.entropia_shannon(dist_antes))
    ))

    print(f"\n{_total_testes - _testes_com_falha}/{_total_testes} testes passaram.")
    return {
        "bruto": bruto, "compativeis": compativeis, "mapeados": mapeados, "completo": completo,
        "dist_antes": dist_antes, "dist_depois": dist_depois,
    }


def _assert(condicao, msg="assercao falhou"):
    if not condicao:
        raise AssertionError(msg)


# -----------------------------------------------------------------------
# Demo
# -----------------------------------------------------------------------

def rodar_demo(caminho_jsonl):
    print("===== Extra: dataset real alternativo (Dolly-15k) rodando contra o pipeline do Modulo 2.2 =====\n")
    r = rodar_testes(caminho_jsonl)
    c = r["completo"]

    print("\n--- Resumo ---")
    print(f"Dataset bruto: {len(r['bruto'])} exemplos reais (databricks-dolly-15k, CC-BY-SA-3.0)")
    print(f"Compativeis com o schema canonico (tem contexto real): {len(r['compativeis'])}")
    print(f"Distribuicao por fonte antes do dedup: {m2.contar_por_fonte(r['mapeados'], CASO)}")
    print(f"Dedup no dataset INTEIRO (comparando instrucao+entrada): {c['dedup']['candidatos_lsh']} candidatos LSH de {c['dedup']['total_forca_bruta']} pares forca-bruta, {len(c['dedup']['pares'])} pares confirmados, {c['itens_removidos']} itens unicos removidos -> {len(c['sem_duplicatas'])} restantes")
    print(f"Balanceado pro alvo: {c['alocacao']}, total {len(c['balanceado'])}")
    ha = m2.entropia_shannon(r["dist_antes"])
    hd = m2.entropia_shannon(r["dist_depois"])
    print(f"Entropia antes/depois do balanceamento: {ha:.4f} -> {hd:.4f} nats")
    print(
        "\nComparar so `entrada` no dedup confundiria pares de PERGUNTA DIFERENTE sobre o MESMO "
        "trecho de contexto (Jaccard=1,0 na entrada) com duplicata real. Comparar instrucao+entrada "
        "evita isso."
    )


if __name__ == "__main__":
    caminho = sys.argv[1] if len(sys.argv) > 1 else os.path.join(_AQUI, "databricks-dolly-15k.jsonl")
    if not os.path.exists(caminho):
        print(f"Arquivo nao encontrado: {caminho}", file=sys.stderr)
        print(
            'Baixe primeiro: curl -L "https://huggingface.co/datasets/databricks/databricks-dolly-15k/resolve/main/databricks-dolly-15k.jsonl" -o databricks-dolly-15k.jsonl',
            file=sys.stderr,
        )
        sys.exit(1)
    rodar_demo(caminho)

# Ahirton Lopes - Fine-Tuning Toolkit - UNIPDS: Processamento de Dados e Fine-Tuning de Modelos
# Prof. Ahirton Lopes, Ph.D. - GDE AI, Microsoft MVP, Senior Manager
